File: ALODWebUtility/Common/PasCode.vb
Imports System.Data.Common
Imports ALOD.Core.Domain.Users
Imports ALOD.Data

Namespace Common

    <Serializable()>
    Public Class pascode

#Region "OldPasCodeDefinition"

        Protected _canEdit As Boolean = False
        Protected _description As String = String.Empty
        Protected _designator As String = String.Empty
        Protected _installation As String = String.Empty
        Protected _majcom As String = String.Empty
        Protected _organization As String = String.Empty
        Protected _orgnKind As String = String.Empty

        'Service Plan Activity
        Protected _parent As String = String.Empty

        Protected _pasCode As String
        Protected _spa As String = String.Empty
        Protected _status As AccessStatus = AccessStatus.Pending
        Protected _unit As String = String.Empty

#Region "PropertiesOld"

        Public Property CanEdit() As Boolean
            Get
                Return _canEdit
            End Get
            Set(ByVal value As Boolean)
                _canEdit = value
            End Set
        End Property

        Public Property Description() As String
            Get
                Return _description
            End Get
            Set(ByVal value As String)
                _description = value
            End Set
        End Property

        Public Property Designator() As String
            Get
                Return _designator
            End Get
            Set(ByVal value As String)
                _designator = value
            End Set
        End Property

        Public Property Installation() As String
            Get
                Return _installation
            End Get
            Set(ByVal value As String)
                _installation = value
            End Set
        End Property

        Public Property MAJCOM() As String
            Get
                Return _majcom
            End Get
            Set(ByVal value As String)
                _majcom = value
            End Set
        End Property

        Public Property Organization() As String
            Get
                Return _organization
            End Get
            Set(ByVal value As String)
                _organization = value
            End Set
        End Property

        Public Property OrgnKind() As String
            Get
                Return _orgnKind
            End Get
            Set(ByVal value As String)
                _orgnKind = value
            End Set
        End Property

        Public Property Parent() As String
            Get
                Return _parent
            End Get
            Set(ByVal value As String)
                _parent = value
            End Set
        End Property

        Public Property SPA() As String
            Get
                Return _spa
            End Get
            Set(ByVal value As String)
                _spa = value
            End Set
        End Property

        Public Property Status() As AccessStatus
            Get
                Return _status
            End Get
            Set(ByVal value As AccessStatus)
                _status = value
            End Set
        End Property

        Public Property Unit() As String
            Get
                Return _unit
            End Get
            Set(ByVal value As String)
                _unit = value
            End Set
        End Property

        Public Property Value() As String
            Get
                Return _pasCode
            End Get
            Set(ByVal value As String)
                If (value.Trim.Length = 8) Then
                    _pasCode = value.Substring(4, 4)
                Else
                    _pasCode = value
                End If
            End Set
        End Property

#End Region

#Region "Constructors"

        Public Sub New()

        End Sub

        Public Sub New(ByVal pCode As pascode)
            _pasCode = pCode.Value
            _description = pCode.Description
            _status = pCode.Status
            _canEdit = pCode.CanEdit

            _spa = pCode.SPA
            _majcom = pCode.MAJCOM
            _designator = pCode.Designator
            _unit = pCode.Unit
            _organization = pCode.Organization  'Service Plan Activity
            _installation = pCode.Installation
            _orgnKind = pCode.OrgnKind
            _parent = pCode.Parent
            _orgnKind = pCode.OrgnKind
            _uic = pCode.uic
        End Sub

        Public Sub New(ByVal pasCode As String)
            _pasCode = pasCode
        End Sub

        Public Sub New(ByVal strpasCode As String, ByVal status As AccessStatus, ByVal canEdit As Boolean)
            _status = status
            _pasCode = strpasCode
            _canEdit = canEdit
        End Sub

        Public Sub New(ByVal strpasCode As String, ByVal status As AccessStatus)
            _status = status
            _pasCode = strpasCode
        End Sub

        Public Function IsValidPascode(ByVal pascode As String) As Boolean
            Dim adapter As New SqlDataStore()
            Dim cmd As System.Data.Common.DbCommand = adapter.GetStoredProcCommand("core_pascode_sp_Validation")
            adapter.AddInParameter(cmd, "@pascode", Data.DbType.String, pascode)
            adapter.AddOutParameter(cmd, "@valid", Data.DbType.Boolean, 0)
            adapter.ExecuteNonQuery(cmd)
            Return CType(cmd.Parameters("@valid").Value, Boolean)

        End Function

#End Region

#End Region

#Region "NewPasCodeDefinition"

#Region "Members"

        Protected _address1 As String = String.Empty

        Protected _address2 As String = String.Empty

        Protected _base_code As String = String.Empty

        Protected _city As String = String.Empty

        Protected _component As String = String.Empty

        Protected _country As String = String.Empty

        '--These are the properties
        Protected _cs_id As Int32

        'These need to be initialized to these values since 0 correspoonds to an existing unit
        Protected _cs_id_parent As Int32 = -1

        Protected _cs_level As String = String.Empty
        Protected _cs_oper_type As String = String.Empty
        Protected _email As String = String.Empty
        Protected _gaining_command_cs_id As Int32 = -1
        Protected _long_name As String = String.Empty
        Protected _pas_code As String = String.Empty
        Protected _postal_code As String = String.Empty
        Protected _reportingStructure As Dictionary(Of String, Int32)
        Protected _state As String = String.Empty
        Protected _time_zone As String = String.Empty
        Protected _uic As String = String.Empty
        Protected _unit_det As String = String.Empty
        Protected _unit_kind As String = String.Empty
        Protected _unit_nbr As String = String.Empty
        Protected _unit_type As String = String.Empty

#End Region

#Region "Properties"

        Public Property address1() As String
            Get
                Return _address1
            End Get
            Set(ByVal value As String)
                _address1 = value
            End Set
        End Property

        Public Property address2() As String
            Get
                Return _address2
            End Get
            Set(ByVal value As String)
                _address2 = value
            End Set
        End Property

        Public Property base_code() As String
            Get
                Return _base_code
            End Get
            Set(ByVal value As String)
                _base_code = value
            End Set
        End Property

        Public Property city() As String
            Get
                Return _city
            End Get
            Set(ByVal value As String)
                _city = value
            End Set
        End Property

        Public Property component() As String
            Get
                Return _component
            End Get
            Set(ByVal value As String)
                _component = value
            End Set
        End Property

        Public Property country() As String
            Get
                Return _country
            End Get
            Set(ByVal value As String)
                _country = value
            End Set
        End Property

        Public Property cs_id() As Integer
            Get
                Return _cs_id
            End Get
            Set(ByVal value As Integer)
                _cs_id = value
            End Set
        End Property

        Public Property cs_id_parent() As Integer
            Get
                Return _cs_id_parent
            End Get
            Set(ByVal value As Integer)
                _cs_id_parent = value
            End Set
        End Property

        Public Property cs_level() As String
            Get
                Return _cs_level
            End Get
            Set(ByVal value As String)
                _cs_level = value
            End Set
        End Property

        Public Property cs_oper_type() As String
            Get
                Return _cs_oper_type
            End Get
            Set(ByVal value As String)
                _cs_oper_type = value
            End Set
        End Property

        Public Property Email() As String
            Get
                Return _email
            End Get
            Set(ByVal value As String)
                _email = value
            End Set
        End Property

        Public Property gaining_command_cs_id() As Integer
            Get
                Return _gaining_command_cs_id
            End Get
            Set(ByVal value As Integer)
                _gaining_command_cs_id = value
            End Set
        End Property

        Public Property long_name() As String
            Get
                Return _long_name
            End Get
            Set(ByVal value As String)
                _long_name = value
            End Set
        End Property

        Public Property pas_code() As String
            Get
                Return _pas_code
            End Get
            Set(ByVal value As String)
                _pas_code = value
            End Set
        End Property

        Public Property postal_code() As String
            Get
                Return _postal_code
            End Get
            Set(ByVal value As String)
                _postal_code = value
            End Set
        End Property

        Public Property reportingStructure() As Dictionary(Of String, Int32)
            Get
                If _reportingStructure Is Nothing Then
                    _reportingStructure = New Dictionary(Of String, Int32)
                End If
                Return _reportingStructure
            End Get
            Set(ByVal value As Dictionary(Of String, Int32))
                _reportingStructure = value
            End Set
        End Property

        Public Property state() As String
            Get
                Return _state
            End Get
            Set(ByVal value As String)
                _state = value
            End Set
        End Property

        Public Property time_zone() As String
            Get
                Return _time_zone
            End Get
            Set(ByVal value As String)
                _time_zone = value
            End Set
        End Property

        Public Property uic() As String
            Get
                Return _uic
            End Get
            Set(ByVal value As String)
                _uic = value
            End Set
        End Property

        Public Property unit_det() As String
            Get
                Return _unit_det
            End Get
            Set(ByVal value As String)
                _unit_det = value
            End Set
        End Property

        Public Property unit_kind() As String
            Get
                Return _unit_kind
            End Get
            Set(ByVal value As String)
                _unit_kind = value
            End Set
        End Property

        Public Property unit_nbr() As String
            Get
                Return _unit_nbr
            End Get
            Set(ByVal value As String)
                _unit_nbr = value
            End Set
        End Property

        Public Property unit_type() As String
            Get
                Return _unit_type
            End Get
            Set(ByVal value As String)
                _unit_type = value
            End Set
        End Property

#End Region

#Region "Constructor"

        Public Sub New(ByVal csid As Integer)
            _cs_id = csid
            _reportingStructure = New Dictionary(Of String, Int32)

        End Sub

#End Region

#Region "Load"

        Public Sub LoadPasCode()

            ' Do not use Iif (evaluates both results [Then and Else] before the conditional [If]) - which causes errors
            Dim unitDao As UnitDao = New NHibernateDaoFactory().GetUnitDao()
            Dim pCodeInfo As Unit = unitDao.GetById(cs_id)
            If (pCodeInfo.Name Is Nothing) Then
                long_name = ""
            Else
                long_name = pCodeInfo.Name
            End If
            If (pCodeInfo.UnitNumber Is Nothing) Then
                unit_nbr = ""
            Else
                unit_nbr = pCodeInfo.UnitNumber
            End If
            If (pCodeInfo.UnitKind Is Nothing) Then
                unit_kind = ""
            Else
                unit_kind = pCodeInfo.UnitKind
            End If
            If (pCodeInfo.UnitType Is Nothing) Then
                unit_type = ""
            Else
                unit_type = pCodeInfo.UnitType
            End If
            If (pCodeInfo.UnitDet Is Nothing) Then
                unit_det = ""
            Else
                unit_det = pCodeInfo.UnitDet
            End If
            If (pCodeInfo.Uic Is Nothing) Then
                uic = ""
            Else
                uic = pCodeInfo.Uic
            End If
            If (pCodeInfo.CommandStructLevel Is Nothing) Then
                cs_level = ""
            Else
                cs_level = pCodeInfo.CommandStructLevel
            End If
            If (pCodeInfo.ParentUnit Is Nothing) Then
                cs_id_parent = -1
            Else
                cs_id_parent = pCodeInfo.ParentUnit.Id
            End If
            If (pCodeInfo.GainingCommand Is Nothing) Then
                gaining_command_cs_id = -1
            Else
                gaining_command_cs_id = pCodeInfo.GainingCommand.Id
            End If
            If (pCodeInfo.PasCode Is Nothing) Then
                pas_code = ""
            Else
                pas_code = pCodeInfo.PasCode
            End If
            If (pCodeInfo.BaseCode Is Nothing) Then
                base_code = ""
            Else
                base_code = pCodeInfo.BaseCode
            End If
            If (pCodeInfo.CommandStructOperationType Is Nothing) Then
                cs_oper_type = ""
            Else
                cs_oper_type = pCodeInfo.CommandStructOperationType
            End If
            If (pCodeInfo.TimeZone Is Nothing) Then
                time_zone = ""
            Else
                time_zone = pCodeInfo.TimeZone
            End If
            If (pCodeInfo.Component Is Nothing) Then
                component = ""
            Else
                component = pCodeInfo.Component
            End If
            If (pCodeInfo.Address1 Is Nothing) Then
                address1 = ""
            Else
                address1 = pCodeInfo.Address1
            End If
            If (pCodeInfo.Address2 Is Nothing) Then
                address2 = ""
            Else
                address2 = pCodeInfo.Address2
            End If
            If (pCodeInfo.City Is Nothing) Then
                city = ""
            Else
                city = pCodeInfo.City
            End If
            If (pCodeInfo.State Is Nothing) Then
                state = ""
            Else
                state = pCodeInfo.State
            End If
            If (pCodeInfo.PostalCode Is Nothing) Then
                postal_code = ""
            Else
                postal_code = pCodeInfo.PostalCode
            End If
            If (pCodeInfo.Country Is Nothing) Then
                country = ""
            Else
                country = pCodeInfo.Country
            End If
            If (pCodeInfo.Email Is Nothing) Then
                Email = ""
            Else
                Email = pCodeInfo.Email
            End If

        End Sub

#End Region

#Region "Update"

        Public Sub UpdatePasCode()

            Dim adapter As New SqlDataStore
            Dim dbCommand As DbCommand

            dbCommand = adapter.GetStoredProcCommand("core_pascode_sp_update")

            adapter.AddInParameter(dbCommand, "@CS_ID", DbType.Int32, cs_id)
            If long_name <> "" Then adapter.AddInParameter(dbCommand, "@LONG_NAME", DbType.String, long_name)
            If unit_nbr <> "" Then adapter.AddInParameter(dbCommand, "@UNIT_NBR", DbType.String, unit_nbr)
            If unit_kind <> "" Then adapter.AddInParameter(dbCommand, "@UNIT_KIND", DbType.String, unit_kind)
            If unit_type <> "" Then adapter.AddInParameter(dbCommand, "@UNIT_TYPE", DbType.String, unit_type)
            If unit_det <> "" Then adapter.AddInParameter(dbCommand, "@UNIT_DET", DbType.String, unit_det)
            If uic <> "" Then adapter.AddInParameter(dbCommand, "@UIC", DbType.String, uic)
            If cs_level <> "" Then adapter.AddInParameter(dbCommand, "@CS_LEVEL", DbType.String, cs_level)
            If cs_id_parent <> -1 Then adapter.AddInParameter(dbCommand, "@CS_ID_PARENT", DbType.Int32, cs_id_parent)
            If gaining_command_cs_id <> -1 Then adapter.AddInParameter(dbCommand, "@GAINING_COMMAND_CS_ID", DbType.Int32, gaining_command_cs_id)
            If pas_code <> "" Then adapter.AddInParameter(dbCommand, "@PAS_CODE", DbType.String, pas_code)
            If base_code <> "" Then adapter.AddInParameter(dbCommand, "@BASE_CODE", DbType.String, base_code)
            If cs_oper_type <> "" Then adapter.AddInParameter(dbCommand, "@CS_OPER_TYPE", DbType.String, cs_oper_type)
            If time_zone <> "" Then adapter.AddInParameter(dbCommand, "@TIME_ZONE", DbType.String, time_zone)
            If component <> "" Then adapter.AddInParameter(dbCommand, "@COMPONENT", DbType.String, component)
            If address1 <> "" Then adapter.AddInParameter(dbCommand, "@ADDRESS1", DbType.String, address1)
            If address2 <> "" Then adapter.AddInParameter(dbCommand, "@ADDRESS2", DbType.String, address2)
            If city <> "" Then adapter.AddInParameter(dbCommand, "@CITY", DbType.String, city)
            If country <> "" Then adapter.AddInParameter(dbCommand, "@COUNTRY", DbType.String, country)
            If state <> "" Then adapter.AddInParameter(dbCommand, "@STATE", DbType.String, state)
            If postal_code <> "" Then adapter.AddInParameter(dbCommand, "@POSTAL_CODE", DbType.String, postal_code)
            If Email <> "" Then adapter.AddInParameter(dbCommand, "@E_MAIL", DbType.String, Email)

            adapter.ExecuteNonQuery(dbCommand)

        End Sub

        Public Sub UpdateReporting(ByVal userId As Integer)

            Dim xml As New XMLString("ReportList")
            For Each entry As KeyValuePair(Of String, Int32) In reportingStructure

                xml.BeginElement("command")
                xml.WriteAttribute("cs_id", cs_id)
                xml.WriteAttribute("chain_type", entry.Key)
                xml.WriteAttribute("parent_cs_id", entry.Value)
                xml.EndElement()

            Next

            Dim adapter As New SqlDataStore
            adapter.ExecuteNonQuery("core_pascode_sp_updateReporting", userId, xml.ToString())

        End Sub

#End Region

#End Region

    End Class

End Namespace

File: ALODWebUtility.CS/Common/PasCode.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Common;
using ALOD.Core.Domain.Users;
using ALOD.Data;

namespace ALODWebUtility.Common
{
    [Serializable]
    public class Pascode
    {
        #region OldPasCodeDefinition

        protected bool _canEdit = false;
        protected string _description = string.Empty;
        protected string _designator = string.Empty;
        protected string _installation = string.Empty;
        protected string _majcom = string.Empty;
        protected string _organization = string.Empty;
        protected string _orgnKind = string.Empty;

        //Service Plan Activity
        protected string _parent = string.Empty;

        protected string _pasCode;
        protected string _spa = string.Empty;
        protected AccessStatus _status = AccessStatus.Pending;
        protected string _unit = string.Empty;

        #region PropertiesOld

        public bool CanEdit
        {
            get { return _canEdit; }
            set { _canEdit = value; }
        }

        public string Description
        {
            get { return _description; }
            set { _description = value; }
        }

        public string Designator
        {
            get { return _designator; }
            set { _designator = value; }
        }

        public string Installation
        {
            get { return _installation; }
            set { _installation = value; }
        }

        public string MAJCOM
        {
            get { return _majcom; }
            set { _majcom = value; }
        }

        public string Organization
        {
            get { return _organization; }
            set { _organization = value; }
        }

        public string OrgnKind
        {
            get { return _orgnKind; }
            set { _orgnKind = value; }
        }

        public string Parent
        {
            get { return _parent; }
            set { _parent = value; }
        }

        public string SPA
        {
            get { return _spa; }
            set { _spa = value; }
        }

        public AccessStatus Status
        {
            get { return _status; }
            set { _status = value; }
        }

        public string Unit
        {
            get { return _unit; }
            set { _unit = value; }
        }

        public string Value
        {
            get { return _pasCode; }
            set
            {
                if (value.Trim().Length == 8)
                {
                    _pasCode = value.Substring(4, 4);
                }
                else
                {
                    _pasCode = value;
                }
            }
        }

        #endregion

        #region Constructors

        public Pascode()
        {
        }

        public Pascode(Pascode pCode)
        {
            _pasCode = pCode.Value;
            _description = pCode.Description;
            _status = pCode.Status;
            _canEdit = pCode.CanEdit;

            _spa = pCode.SPA;
            _majcom = pCode.MAJCOM;
            _designator = pCode.Designator;
            _unit = pCode.Unit;
            _organization = pCode.Organization;  //Service Plan Activity
            _installation = pCode.Installation;
            _orgnKind = pCode.OrgnKind;
            _parent = pCode.Parent;
            _orgnKind = pCode.OrgnKind;
            _uic = pCode.uic;
        }

        public Pascode(string pasCode)
        {
            _pasCode = pasCode;
        }

        public Pascode(string strpasCode, AccessStatus status, bool canEdit)
        {
            _status = status;
            _pasCode = strpasCode;
            _canEdit = canEdit;
        }

        public Pascode(string strpasCode, AccessStatus status)
        {
            _status = status;
            _pasCode = strpasCode;
        }

        public bool IsValidPascode(string pascode)
        {
            SqlDataStore adapter = new SqlDataStore();
            DbCommand cmd = adapter.GetStoredProcCommand("core_pascode_sp_Validation");
            adapter.AddInParameter(cmd, "@pascode", DbType.String, pascode);
            adapter.AddOutParameter(cmd, "@valid", DbType.Boolean, 0);
            adapter.ExecuteNonQuery(cmd);
            return Convert.ToBoolean(cmd.Parameters["@valid"].Value);
        }

        #endregion

        #endregion

        #region NewPasCodeDefinition

        #region Members

        protected string _address1 = string.Empty;

        protected string _address2 = string.Empty;

        protected string _base_code = string.Empty;

        protected string _city = string.Empty;

        protected string _component = string.Empty;

        protected string _country = string.Empty;

        //--These are the properties
        protected int _cs_id;

        //These need to be initialized to these values since 0 correspoonds to an existing unit
        protected int _cs_id_parent = -1;

        protected string _cs_level = string.Empty;
        protected string _cs_oper_type = string.Empty;
        protected string _email = string.Empty;
        protected int _gaining_command_cs_id = -1;
        protected string _long_name = string.Empty;
        protected string _pas_code = string.Empty;
        protected string _postal_code = string.Empty;
        protected Dictionary<string, int> _reportingStructure;
        protected string _state = string.Empty;
        protected string _time_zone = string.Empty;
        protected string _uic = string.Empty;
        protected string _unit_det = string.Empty;
        protected string _unit_kind = string.Empty;
        protected string _unit_nbr = string.Empty;
        protected string _unit_type = string.Empty;

        #endregion

        #region Properties

        public string address1
        {
            get { return _address1; }
            set { _address1 = value; }
        }

        public string address2
        {
            get { return _address2; }
            set { _address2 = value; }
        }

        public string base_code
        {
            get { return _base_code; }
            set { _base_code = value; }
        }

        public string city
        {
            get { return _city; }
            set { _city = value; }
        }

        public string component
        {
            get { return _component; }
            set { _component = value; }
        }

        public string country
        {
            get { return _country; }
            set { _country = value; }
        }

        public int cs_id
        {
            get { return _cs_id; }
            set { _cs_id = value; }
        }

        public int cs_id_parent
        {
            get { return _cs_id_parent; }
            set { _cs_id_parent = value; }
        }

        public string cs_level
        {
            get { return _cs_level; }
            set { _cs_level = value; }
        }

        public string cs_oper_type
        {
            get { return _cs_oper_type; }
            set { _cs_oper_type = value; }
        }

        public string Email
        {
            get { return _email; }
            set { _email = value; }
        }

        public int gaining_command_cs_id
        {
            get { return _gaining_command_cs_id; }
            set { _gaining_command_cs_id = value; }
        }

        public string long_name
        {
            get { return _long_name; }
            set { _long_name = value; }
        }

        public string pas_code
        {
            get { return _pas_code; }
            set { _pas_code = value; }
        }

        public string postal_code
        {
            get { return _postal_code; }
            set { _postal_code = value; }
        }

        public Dictionary<string, int> reportingStructure
        {
            get
            {
                if (_reportingStructure == null)
                {
                    _reportingStructure = new Dictionary<string, int>();
                }
                return _reportingStructure;
            }
            set { _reportingStructure = value; }
        }

        public string state
        {
            get { return _state; }
            set { _state = value; }
        }

        public string time_zone
        {
            get { return _time_zone; }
            set { _time_zone = value; }
        }

        public string uic
        {
            get { return _uic; }
            set { _uic = value; }
        }

        public string unit_det
        {
            get { return _unit_det; }
            set { _unit_det = value; }
        }

        public string unit_kind
        {
            get { return _unit_kind; }
            set { _unit_kind = value; }
        }

        public string unit_nbr
        {
            get { return _unit_nbr; }
            set { _unit_nbr = value; }
        }

        public string unit_type
        {
            get { return _unit_type; }
            set { _unit_type = value; }
        }

        #endregion

        #region Constructor

        public Pascode(int csid)
        {
            _cs_id = csid;
            _reportingStructure = new Dictionary<string, int>();
        }

        #endregion

        #region Load

        public void LoadPasCode()
        {
            // Do not use Iif (evaluates both results [Then and Else] before the conditional [If]) - which causes errors
            UnitDao unitDao = new NHibernateDaoFactory().GetUnitDao();
            Unit pCodeInfo = unitDao.GetById(cs_id);
            if (pCodeInfo.Name == null)
            {
                long_name = "";
            }
            else
            {
                long_name = pCodeInfo.Name;
            }
            if (pCodeInfo.UnitNumber == null)
            {
                unit_nbr = "";
            }
            else
            {
                unit_nbr = pCodeInfo.UnitNumber;
            }
            if (pCodeInfo.UnitKind == null)
            {
                unit_kind = "";
            }
            else
            {
                unit_kind = pCodeInfo.UnitKind;
            }
            if (pCodeInfo.UnitType == null)
            {
                unit_type = "";
            }
            else
            {
                unit_type = pCodeInfo.UnitType;
            }
            if (pCodeInfo.UnitDet == null)
            {
                unit_det = "";
            }
            else
            {
                unit_det = pCodeInfo.UnitDet;
            }
            if (pCodeInfo.Uic == null)
            {
                uic = "";
            }
            else
            {
                uic = pCodeInfo.Uic;
            }
            if (pCodeInfo.CommandStructLevel == null)
            {
                cs_level = "";
            }
            else
            {
                cs_level = pCodeInfo.CommandStructLevel;
            }
            if (pCodeInfo.ParentUnit == null)
            {
                cs_id_parent = -1;
            }
            else
            {
                cs_id_parent = pCodeInfo.ParentUnit.Id;
            }
            if (pCodeInfo.GainingCommand == null)
            {
                gaining_command_cs_id = -1;
            }
            else
            {
                gaining_command_cs_id = pCodeInfo.GainingCommand.Id;
            }
            if (pCodeInfo.PasCode == null)
            {
                pas_code = "";
            }
            else
            {
                pas_code = pCodeInfo.PasCode;
            }
            if (pCodeInfo.BaseCode == null)
            {
                base_code = "";
            }
            else
            {
                base_code = pCodeInfo.BaseCode;
            }
            if (pCodeInfo.CommandStructOperationType == null)
            {
                cs_oper_type = "";
            }
            else
            {
                cs_oper_type = pCodeInfo.CommandStructOperationType;
            }
            if (pCodeInfo.TimeZone == null)
            {
                time_zone = "";
            }
            else
            {
                time_zone = pCodeInfo.TimeZone;
            }
            if (pCodeInfo.Component == null)
            {
                component = "";
            }
            else
            {
                component = pCodeInfo.Component;
            }
            if (pCodeInfo.Address1 == null)
            {
                address1 = "";
            }
            else
            {
                address1 = pCodeInfo.Address1;
            }
            if (pCodeInfo.Address2 == null)
            {
                address2 = "";
            }
            else
            {
                address2 = pCodeInfo.Address2;
            }
            if (pCodeInfo.City == null)
            {
                city = "";
            }
            else
            {
                city = pCodeInfo.City;
            }
            if (pCodeInfo.State == null)
            {
                state = "";
            }
            else
            {
                state = pCodeInfo.State;
            }
            if (pCodeInfo.PostalCode == null)
            {
                postal_code = "";
            }
            else
            {
                postal_code = pCodeInfo.PostalCode;
            }
            if (pCodeInfo.Country == null)
            {
                country = "";
            }
            else
            {
                country = pCodeInfo.Country;
            }
            if (pCodeInfo.Email == null)
            {
                Email = "";
            }
            else
            {
                Email = pCodeInfo.Email;
            }

        }

        #endregion

        #region Update

        public void UpdatePasCode()
        {

            SqlDataStore adapter = new SqlDataStore();
            DbCommand dbCommand;

            dbCommand = adapter.GetStoredProcCommand("core_pascode_sp_update");

            adapter.AddInParameter(dbCommand, "@CS_ID", DbType.Int32, cs_id);
            if (long_name != "") adapter.AddInParameter(dbCommand, "@LONG_NAME", DbType.String, long_name);
            if (unit_nbr != "") adapter.AddInParameter(dbCommand, "@UNIT_NBR", DbType.String, unit_nbr);
            if (unit_kind != "") adapter.AddInParameter(dbCommand, "@UNIT_KIND", DbType.String, unit_kind);
            if (unit_type != "") adapter.AddInParameter(dbCommand, "@UNIT_TYPE", DbType.String, unit_type);
            if (unit_det != "") adapter.AddInParameter(dbCommand, "@UNIT_DET", DbType.String, unit_det);
            if (uic != "") adapter.AddInParameter(dbCommand, "@UIC", DbType.String, uic);
            if (cs_level != "") adapter.AddInParameter(dbCommand, "@CS_LEVEL", DbType.String, cs_level);
            if (cs_id_parent != -1) adapter.AddInParameter(dbCommand, "@CS_ID_PARENT", DbType.Int32, cs_id_parent);
            if (gaining_command_cs_id != -1) adapter.AddInParameter(dbCommand, "@GAINING_COMMAND_CS_ID", DbType.Int32, gaining_command_cs_id);
            if (pas_code != "") adapter.AddInParameter(dbCommand, "@PAS_CODE", DbType.String, pas_code);
            if (base_code != "") adapter.AddInParameter(dbCommand, "@BASE_CODE", DbType.String, base_code);
            if (cs_oper_type != "") adapter.AddInParameter(dbCommand, "@CS_OPER_TYPE", DbType.String, cs_oper_type);
            if (time_zone != "") adapter.AddInParameter(dbCommand, "@TIME_ZONE", DbType.String, time_zone);
            if (component != "") adapter.AddInParameter(dbCommand, "@COMPONENT", DbType.String, component);
            if (address1 != "") adapter.AddInParameter(dbCommand, "@ADDRESS1", DbType.String, address1);
            if (address2 != "") adapter.AddInParameter(dbCommand, "@ADDRESS2", DbType.String, address2);
            if (city != "") adapter.AddInParameter(dbCommand, "@CITY", DbType.String, city);
            if (country != "") adapter.AddInParameter(dbCommand, "@COUNTRY", DbType.String, country);
            if (state != "") adapter.AddInParameter(dbCommand, "@STATE", DbType.String, state);
            if (postal_code != "") adapter.AddInParameter(dbCommand, "@POSTAL_CODE", DbType.String, postal_code);
            if (Email != "") adapter.AddInParameter(dbCommand, "@E_MAIL", DbType.String, Email);

            adapter.ExecuteNonQuery(dbCommand);

        }

        public void UpdateReporting(int userId)
        {
            XMLString xml = new XMLString("ReportList");
            foreach (KeyValuePair<string, int> entry in reportingStructure)
            {
                xml.BeginElement("command");
                xml.WriteAttribute("cs_id", cs_id);
                xml.WriteAttribute("chain_type", entry.Key);
                xml.WriteAttribute("parent_cs_id", entry.Value);
                xml.EndElement();
            }

            SqlDataStore adapter = new SqlDataStore();
            adapter.ExecuteNonQuery("core_pascode_sp_updateReporting", userId, xml.ToString());
        }

        #endregion

        #endregion
    }
}
